### DISCUSSION TITLE
C ++ HSPY File Writer

### DISCUSSION THREAD

**Discussion by CSSFrancis** [hyperspy/rosettasciio#325] [Development]
Direct Electron has a (partial) c++ implementation for writing to the `.hspy` file format as well as for the DE5 file format (a varient of the `.emd` file format).  Something that I would like is to improve the ability for companies to write directly into the `.hspy` file format as well as to the `.emd` file format (and eventually into the `.zspy` file format). 

As far as performance goes, I don't know if there is much difference from C++ --> python but I think in many instances a c++ implimentation should be better able to integrate with different aquisition softwares.

I'd be curious to hear peoples thoughts on the subject? 

1. Is there interest from other companies to write directly into a `.hspy`/ `.zspy` / `emd` file?
2. Are 2 seperate implementations worse than 1 in this case as there could potentially be some conflicts?
3. What liscense should this be released under?  We don't necessarily want to put restrictions on this but it would be nice to know if someone else is using this code comercially. 
4. @sk1p Should we be using Rust or something other than c++ here :).

**Comment by jlaehne** [comment]
@noemiebonnet @pieleric @Attolight-NTappy a discussion possibly relevant for your companies as you at least have reader implementations for RosettaSciIO/HyperSpy

**Comment by sk1p** [comment]
Thanks for tagging me!

> Direct Electron has a (partial) c++ implementation for writing to the `.hspy` file format as well as for the DE5 file format (a varient of the `.emd` file format). Something that I would like is to improve the ability for companies to write directly into the `.hspy` file format as well as to the `.emd` file format (and eventually into the `.zspy` file format).
> 
> As far as performance goes, I don't know if there is much difference from C++ --> python but I think in many instances a c++ implimentation should be better able to integrate with different aquisition softwares.

Indeed, also if you need something resembling (firm) real-time assurances, this is a lot easier with C++/Rust than Python.

>     1. Is there interest from other companies to write directly into a `.hspy`/ `.zspy` / `emd` file?
> 
>     2. Are 2 seperate implementations worse than 1 in this case as there could potentially be some conflicts?

I'm working on adding zarr/zspy support to LiberTEM-live, meaning any detector detector we support will be able to directly produce zspy soon. So that would be implementation number three.

I think it should be easy enough to support the general structure of zspy files, from my perspective. It could be helpful to have a small acceptance test that checks the structure, basically an executable version of the textual description, which might catch subtleties that are not expressed in the language here:

https://github.com/hyperspy/rosettasciio/blob/3c7b9f61845ed304b6e6abb3e7057b983e4e3d2f/rsciio/zspy/_api.py#L42-L67

I think the more challenging part would be writing the metadata tree and keeping that consistent between implementations. See also this [previous discussion](https://github.com/hyperspy/hyperspy/discussions/3431). I think most helpful would be to have a formal source of truth, like a schema, that documents the metadata attributes in a clear way.

>  it would be nice to know if someone else is using this code comercially. 

I think the options are BSD/MIT where you would get a notice in the docs or somewhere in a menu of the application. (A)GPL might be a hard sell to most commercial users, from my experience.

>     4. @sk1p Should we be using Rust or something other than c++ here :).

If there is interest, I would be happy to collaborate on a rust library, with a C or C++ API (and Python?) on top for integration into other software. The argument for rust would be that it's IMHO easier to write correct rust code compared to C++, and with PyO3 there is a really good option for building Python bindings.

**Comment by Attolight-NTappy** [comment]
Thanks for tagging me in this discussion!

We really like the `*.hspy` file format at attolight and recommend using it to our customers for sharing data and saving them once processed using hyperspy. 

In the future, there could be use cases where writing data to that format directly would be interesting. However, I am not certain that licensing conditions (also of hdf5 data standard) allow to do this easily.

**Reply by sk1p** [reply]
> also of hdf5 data standard

AFAIK the libraries (C, Python) are BSD licensed, so compliance should be easy even for closed-source software
